espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ComposeNodeFinder.kt

```diff
@@ -0,0 +1,80 @@
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.helpers
+
+import androidx.compose.ui.test.*
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidSelectorException
+import io.appium.espressoserver.lib.model.Strategy
+import io.appium.espressoserver.EspressoServerRunnerTest
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException
+import io.appium.espressoserver.lib.handlers.exceptions.StaleElementException
+import io.appium.espressoserver.lib.model.Locator
+import java.util.NoSuchElementException
+
+/**
+ * Retrieve cached node and return the SemanticsNodeInteraction
+ */
+fun getNodeInteractionById(elementId: String?): SemanticsNodeInteraction =
+    elementId?.let { ComposeViewCache.get(it) ?: throw StaleElementException(it) }
+        ?: throw InvalidArgumentException("Cannot find 'null' element")
+
+
+/**
+ * Retrieve cached SemanticsNodeInteraction and ensure its not stale
+ */
+fun getValidNodeInteractionById(elementId: String?): SemanticsNodeInteraction =
+    try {
+        val nodeInteraction = getNodeInteractionById(elementId)
+        nodeInteraction.assertExists().let { return nodeInteraction }
+    } catch (e: AssertionError) {
+        throw StaleElementException(elementId!!)
+    }
+
+/**
+ * Helper methods to find SemanticsNodeInteractions based on locator strategies and selectors
+ */
+fun findAllComposeNodeInteractions(
+    parentNode: SemanticsNodeInteraction?,
+    locator: Locator
+): SemanticsNodeInteractionCollection {
+    val matcher =
+        when (locator.using) {
+            Strategy.VIEW_TAG -> hasTestTag(locator.value!!)
+            Strategy.TEXT -> hasText(locator.value!!)
+            Strategy.LINK_TEXT -> hasText(locator.value!!)
+            Strategy.ACCESSIBILITY_ID -> hasContentDescription(locator.value!!)
+            else -> throw InvalidSelectorException(
+                "Can't use non-Compose selectors. " +
+                        "only ${Strategy.VIEW_TAG}, ${Strategy.TEXT}, ${Strategy.LINK_TEXT} and " +
+                        "${Strategy.ACCESSIBILITY_ID} are supported"
+            )
+        }
+
+    return if (parentNode == null) {
+        EspressoServerRunnerTest.composeTestRule.onAllNodes(matcher)
+    } else {
+        parentNode.onChildren().filter(matcher)
+    }
+}
+
+fun getSemanticsNode(elementId: String) =
```

Comment:
I prefer to always provide returned function type explicitly, so the code is easier to read

Reply:
added